sw/device/tests/sim_dv/sram_ctrl_scrambled_access_test.c

```diff
@@ -0,0 +1,374 @@
+  memcpy(kBackdoorExpectedWords, kBackdoorExpectedBytes, kTestBufferSizeBytes);
+
+  CHECK_ARRAYS_EQ(mem_frame->backdoor, kBackdoorExpectedWords,
+                  kTestBufferSizeWords);
+}
+
+static void execute_retention_sram_test(void) {
+  LOG_INFO("Wiping retention sram...");
+  CHECK_DIF_OK(dif_sram_ctrl_wipe(&ret_sram));
+
+  LOG_INFO("Preparing test...");
+  prepare_sram_for_scrambling((uintptr_t)reference_frame->pattern,
+                              (uintptr_t)scrambling_frame->pattern);
+
+  LOG_INFO("Scrambling...");
+  sram_ctrl_testutils_scramble(&ret_sram);
+
+  // Reset the Ecc error count that lies on the main sram.
+  LOG_INFO("Checking memory...");
+  reference_frame->ecc_error_counter = 0;
+  check_sram_data(scrambling_frame);
+}
+
+/**
+ * Override internal IRQ interrupt service routine to count
+ * the number of integrity exceptions.
+ */
+void ottf_internal_isr(void) {
+  LOG_INFO("%s - %d", __func__, reference_frame->ecc_error_counter);
+  reference_frame->ecc_error_counter++;
+}
+
+static void sync_testbench(void) {
+  // Set WFI status for testbench synchronization,
+  // no actual WFI instruction is issued.
+  test_status_set(kTestStatusInWfi);
+  test_status_set(kTestStatusInTest);
+}
+
+/**
+ * Executes the MAIN SRAM scrambling test.
+ *
+ * This test is re-entrant. On the first pass the test triggers MAIN SRAM
+ * scrambling, saves the relevant test data in RETENTION SRAM, and triggers the
+ * system reset. On the second pass, it compares the saved test data against
+ * the expected pattern.
+ */
+uint32_t main_sram_addr;
+uint32_t ret_sram_addr;
+
+extern uint8_t _stack_start[];
+extern uint8_t _freertos_heap_start[];
+
+bool test_main(void) {
+  CHECK_DIF_OK(dif_rstmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_RSTMGR_AON_BASE_ADDR), &rstmgr));
+
+  CHECK_DIF_OK(dif_sram_ctrl_init(
+      mmio_region_from_addr(TOP_EARLGREY_SRAM_CTRL_RET_AON_REGS_BASE_ADDR),
+      &ret_sram));
+
+  main_sram_addr = OT_ALIGN_MEM(rand_testutils_gen32_range(
+      (uintptr_t)_freertos_heap_start,
+      (uintptr_t)_stack_start - sizeof(scramble_test_frame)));
+  ret_sram_addr = OT_ALIGN_MEM(rand_testutils_gen32_range(
+      kRetSramBaseAddr, kRetRamLastAddr - sizeof(scramble_test_frame)));
+
+  scrambling_frame = (scramble_test_frame *)main_sram_addr;
+  reference_frame = (scramble_test_frame *)ret_sram_addr;
+  LOG_INFO("RET_SRAM addr: %x MAIN_SRAM addr: %x", ret_sram_addr,
+           main_sram_addr);
+
+  dif_rstmgr_reset_info_bitfield_t info = rstmgr_testutils_reason_get();
+  if (info == kDifRstmgrResetInfoPor) {
+    sync_testbench();
+    LOG_INFO("First boot, testing main sram");
+    // First boot, start with ret sram.
+    execute_main_sram_test();
+
+  } else if (info == kDifRstmgrResetInfoSw) {
+    LOG_INFO("Second boot, checking main sram");
+
+    check_sram_data(reference_frame);
+
+    // CHECK_ARRAYS_NE(scrambling_frame->pattern, (uint32_t *)kRamTestPattern1,
```

Comment:
I'm still trying to figure out why the first word of the main sram still matches with `kRamTestPattern1` even after the main sram scrambling.